core/src/main/java/io/grpc/internal/TransportSet2.java

```diff
@@ -291,39 +261,77 @@ public void run() {
     }
   }
 
-  @Override
-  public ManagedChannel shutdown() {
+  @GuardedBy("lock")
+  private void gotoNonErrorState(ConnectivityState newState) {
+    gotoState(ConnectivityStateInfo.forNonError(newState));
+  }
+
+  @GuardedBy("lock")
+  private void gotoState(final ConnectivityStateInfo newState) {
+    if (state.getState() != newState.getState()) {
+      if (state.getState() == SHUTDOWN) {
+        throw new IllegalStateException("Cannot transition out of SHUTDOWN to " + newState);
+      }
+      state = newState;
+      channelExecutor.execute(new Runnable() {
+          @Override
+          public void run() {
+            callback.onStateChange(TransportSet2.this, newState);
+          }
+        });
+    }
+  }
+
+  public void shutdown() {
     ManagedClientTransport savedActiveTransport;
     ConnectionClientTransport savedPendingTransport;
-    boolean runCallback = false;
+    boolean terminated = false;
     synchronized (lock) {
-      if (shutdown) {
-        return this;
+      if (state.getState() == SHUTDOWN) {
+        return;
       }
-      stateManager.gotoState(ConnectivityState.SHUTDOWN);
-      shutdown = true;
+      gotoNonErrorState(SHUTDOWN);
       savedActiveTransport = activeTransport;
       savedPendingTransport = pendingTransport;
       activeTransport = null;
+      pendingTransport = null;
       if (transports.isEmpty()) {
-        runCallback = true;
-        terminatedLatch.countDown();
+        terminated = true;
```

Comment:
Just call `handleTermination` here? Ditto in `transportTerminated`? I'm still digging into the threading, but it seems like running the executor is assumed to be safe in `gotoState()`.

Honestly, I don't care too much whether it stays the way it is, but it looked like the use of `terminated` was just because of the history of the code.